File: solutions.digamma.damas.jcr/src/test/java/solutions/digamma/damas/jcr/content/JcrFolderManagerTest.java
package solutions.digamma.damas.jcr.content;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import solutions.digamma.damas.NotFoundException;
import solutions.digamma.damas.auth.Token;
import solutions.digamma.damas.content.Folder;
import solutions.digamma.damas.content.FolderManager;
import solutions.digamma.damas.jcr.Mocks;
import solutions.digamma.damas.jcr.WeldTest;

/**
 * Created by ahmad on 9/3/17.
 */
public class JcrFolderManagerTest extends WeldTest {

    private FolderManager manager;
    private Token token;

    @Before
    public void setUp() throws Exception {
        this.manager = inject(JcrFolderManager.class);
        this.token = this.login.login("admin", "admin");
    }

    @After
    public void tearDown() throws Exception {
        this.login.logout(this.token);
    }

    @Test
    public void retrieve() throws Exception {
        final String name = "test";
        String rootId = this.manager
                .find(this.token).getObjects().iterator().next().getId();
        String id =
                manager.create(this.token, Mocks.folder(rootId, name)).getId();
        Folder folder = manager.retrieve(this.token, id);
        assert name.equals(folder.getName()) : "Folder name mismatch";
        assert rootId.equals(folder.getParentId()) : "Folder ID mismatch";
    }

    @Test
    public void create() throws Exception {
        final String name = "test";
        String rootId = this.manager
                .find(this.token).getObjects().iterator().next().getId();
        Folder folder =
                manager.create(this.token, Mocks.folder(rootId, name));
        assert name.equals(folder.getName()) : "Folder name mismatch";
        assert rootId.equals(folder.getParentId()) : "Folder ID mismatch";
    }

    @Test
    public void updateName() throws Exception {
        String name = "test";
        String rootId = this.manager
                .find(this.token).getObjects().iterator().next().getId();
        String id =
                manager.create(this.token, Mocks.folder(rootId, name)).getId();
        name = "new.txt";
        manager.update(this.token, id, Mocks.folder(null, name));
        Folder folder = manager.retrieve(this.token, id);
        assert id.equals(folder.getId()) : "Updated folder ID mismatch.";
        assert name.equals(folder.getName()) :
                "Updated folder name mismatch.";
        assert rootId.equals(folder.getParentId()) :
                "Updated folder parent ID mismatch.";
    }

    @Test
    public void updateParent() throws Exception {
        String name = "test";
        String rootId = this.manager
                .find(this.token).getObjects().iterator().next().getId();
        String id =
                manager.create(this.token, Mocks.folder(rootId, name)).getId();
        String parentId = manager
                .create(this.token, Mocks.folder(rootId, "parent"))
                .getId();
        manager.update(this.token, id, Mocks.folder(parentId, null));
        Folder folder = manager.retrieve(this.token, id);
        assert id.equals(folder.getId()) : "Updated folder ID mismatch.";
        assert name.equals(folder.getName()) :
                "Updated folder name mismatch.";
        assert parentId.equals(folder.getParentId()) :
                "Updated folder parent ID mismatch.";
    }

    @Test
    public void updateNameAndParent() throws Exception {
        String name = "test";
        String rootId = this.manager
                .find(this.token).getObjects().iterator().next().getId();
        String id =
                manager.create(this.token, Mocks.folder(rootId, name)).getId();
        String parentId = manager
                .create(this.token, Mocks.folder(rootId, "parent"))
                .getId();
        name = "new";
        manager.update(this.token, id, Mocks.folder(parentId, name));
        Folder folder = manager.retrieve(this.token, id);
        assert id.equals(folder.getId()) : "Updated folder ID mismatch.";
        assert name.equals(folder.getName()) :
                "Updated folder name mismatch.";
        assert parentId.equals(folder.getParentId()) :
                "Updated folder parent ID mismatch.";
    }


    @Test
    public void delete() throws Exception {
        String name = "test";
        String rootId = this.manager
                .find(this.token).getObjects().iterator().next().getId();
        String id =
                manager.create(this.token, Mocks.folder(rootId, name)).getId();
        manager.delete(this.token, id);
        try {
            manager.retrieve(this.token, id);
            assert false : "NotFoundException should be thrown.";
        } catch (NotFoundException e) {}
    }

    @Test
    public void find() throws Exception {
    }
}

File: solutions.digamma.damas.jcr/src/test/kotlin/solutions/digamma/damas/jcr/content/JcrFolderManagerTest.kt
package solutions.digamma.damas.jcr.content

import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import solutions.digamma.damas.common.NotFoundException
import solutions.digamma.damas.jcr.Mocks
import solutions.digamma.damas.jcr.WeldTest
import solutions.digamma.damas.search.Filter

/**
 * Test folder manager.
 *
 * Created by Ahmad on 9/3/17.
 */
class JcrFolderManagerTest : WeldTest() {

    private var manager = WeldTest.inject(JcrFolderManager::class.java)

    @Before
    @Throws(Exception::class)
    fun setUp() {
        this.login()
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        this.logout()
    }

    @Test
    @Throws(Exception::class)
    fun retrieve() {
        val name = "test"
        val rootId = this.manager.find("/").id
        val id = manager.create(Mocks.folder(rootId, name)).id
        val folder = manager.retrieve(id)
        assert(name == folder.name) { "Folder name mismatch" }
        assert(rootId == folder.parentId) { "Folder ID mismatch" }
        manager.delete(id)
    }

    @Test
    @Throws(Exception::class)
    fun create() {
        val name = "test"
        val rootId = this.manager.find("/").id
        val folder = manager.create(Mocks.folder(rootId, name))
        assert(name == folder.name) { "Folder name mismatch" }
        assert(rootId == folder.parentId) { "Folder ID mismatch" }
        manager.delete(folder.id)
    }

    @Test
    @Throws(Exception::class)
    fun createWithWhiteSpace() {
        val name = "test folder"
        val rootId = this.manager.find("/").id
        val folder = manager.create(Mocks.folder(rootId, name))
        assert(name == folder.name) { "Folder name mismatch" }
        assert(rootId == folder.parentId) { "Folder ID mismatch" }
        manager.delete(folder.id)
    }

    @Test
    @Throws(Exception::class)
    fun updateName() {
        var name = "test"
        val rootId = this.manager.find("/").id
        val id = manager.create(Mocks.folder(rootId, name)).id
        name = "new.txt"
        manager.update(id, Mocks.folder(null, name))
        val folder = manager.retrieve(id)
        assert(id == folder.id) { "Updated folder ID mismatch." }
        assert(name == folder.name) { "Updated folder name mismatch." }
        assert(rootId == folder.parentId) {
            "Updated folder parent ID mismatch."
        }
        manager.delete(id)
    }

    @Test
    @Throws(Exception::class)
    fun updateNameWithWhiteSpace() {
        var name = "test"
        val rootId = this.manager.find("/").id
        val id = manager.create(Mocks.folder(rootId, name)).id
        name = "new folder.txt"
        manager.update(id, Mocks.folder(null, name))
        val folder = manager.retrieve(id)
        assert(id == folder.id) { "Updated folder ID mismatch." }
        assert(name == folder.name) { "Updated folder name mismatch." }
        assert(rootId == folder.parentId) {
            "Updated folder parent ID mismatch."
        }
        manager.delete(id)
    }

    @Test
    @Throws(Exception::class)
    fun updateParent() {
        val name = "test"
        val rootId = this.manager.find("/").id
        val id = manager.create(Mocks.folder(rootId, name)).id
        val parentId = manager.create(Mocks.folder(rootId, "parent")).id
        manager.update(id, Mocks.folder(parentId, null))
        val folder = manager.retrieve(id)
        assert(id == folder.id) { "Updated folder ID mismatch." }
        assert(name == folder.name) { "Updated folder name mismatch." }
        assert(parentId == folder.parentId) {
            "Updated folder parent ID mismatch."
        }
        manager.delete(id)
        manager.delete(parentId)
    }

    @Test
    @Throws(Exception::class)
    fun updateNameAndParent() {
        var name = "test"
        val rootId = this.manager.find("/").id
        val id = manager.create(Mocks.folder(rootId, name)).id
        val parentId = manager.create(Mocks.folder(rootId, "parent")).id
        name = "new"
        manager.update(id, Mocks.folder(parentId, name))
        val folder = manager.retrieve(id)
        assert(id == folder.id) { "Updated folder ID mismatch." }
        assert(name == folder.name) { "Updated folder name mismatch." }
        assert(parentId == folder.parentId) {
            "Updated folder parent ID mismatch."
        }
        manager.delete(id)
        manager.delete(parentId)
    }

    @Test
    @Throws(Exception::class)
    fun copy() {
        val name = "test"
        val rootId = this.manager.find("/").id
        val id = manager.create(Mocks.folder(rootId, name)).id
        val parentId = manager.create(Mocks.folder(rootId, "parent")).id
        this.commit()
        val copyId = manager.copy(id, parentId).id
        val folder = manager.retrieve(copyId)
        assert(parentId == folder.parentId) { "Copied folder parent mismatch." }
        assert(name == folder.name) { "Copied folder name mismatch." }
        manager.delete(id)
        manager.delete(parentId)
    }


    @Test
    @Throws(Exception::class)
    fun delete() {
        val name = "test"
        val rootId = this.manager.find("/").id
        val id = manager.create(Mocks.folder(rootId, name)).id
        manager.delete(id)
        try {
            manager.retrieve(id)
            assert(false) { "NotFoundException should be thrown." }
        } catch (e: NotFoundException) {
            "${e.javaClass} thrown instead of NotFoundException."
        }
    }

    @Test
    @Throws(Exception::class)
    fun find() {
    }

    @Test
    fun findByName() {
        val name = "test"
        val rootId = this.manager.find("/").id
        val id = manager.create(Mocks.folder(rootId, name)).id
        this.commit()
        val filter = Mockito.mock(Filter::class.java)
        Mockito.`when`(filter.namePattern).thenReturn("te*")
        var page = this.manager.find(0, 30, filter)
        assert(page.total == 1)
        assert(page.size == 1)
        assert(page.objects.size == 1)
        val entity = page.objects.iterator().next()
        assert(name == entity.name)
        Mockito.`when`(filter.namePattern).thenReturn("toast*")
        page = this.manager.find(0, 30, filter)
        assert(page.total == 0)
        assert(page.size == 0)
        assert(page.objects.size == 0)
        this.manager.delete(id)
    }
}
